From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: World of Warcraft <warcraft@users.noreply.github.com>
Date: Fri, 16 Oct 2026 06:16:40 +0000
Subject: [PATCH 1/7] G3D: build for arm64 and musl

Its atomics and cycle counter were x86 assembly, cpuid and <xmmintrin.h> x86 only: the compiler's
atomics, arm64's virtual counter (cntvct_el0), and no cpuid off x86, as TrinityCore's G3D has
since. System.h uses struct timeval without <sys/time.h>; FileSystem.cpp uses stat64, which musl
doesn't have (its stat is 64-bit).
---
 dep/g3dlite/include/G3D/AtomicInt32.h | 26 ++++----------------------
 dep/g3dlite/include/G3D/System.h      | 13 +++++++++++++
 dep/g3dlite/source/FileSystem.cpp     |  4 ++++
 dep/g3dlite/source/System.cpp         |  8 +++++++-
 4 files changed, 28 insertions(+), 23 deletions(-)

diff --git a/dep/g3dlite/include/G3D/AtomicInt32.h b/dep/g3dlite/include/G3D/AtomicInt32.h
index 7b56e00..6441192 100644
--- a/dep/g3dlite/include/G3D/AtomicInt32.h
+++ b/dep/g3dlite/include/G3D/AtomicInt32.h
@@ -76,12 +76,8 @@ public:
 
 #       elif defined(G3D_LINUX) || defined(G3D_FREEBSD)
 
-            int32 old;
-            asm volatile ("lock; xaddl %0,%1"
-                  : "=r"(old), "=m"(m_value) /* outputs */
-                  : "0"(x), "m"(m_value)   /* inputs */
-                  : "memory", "cc");
-            return old;
+            // The compiler's atomics: the x86 assembly here before doesn't build for arm64.
+            return __sync_fetch_and_add(&m_value, x);
             
 #       elif defined(G3D_OSX)
 
@@ -115,14 +111,7 @@ public:
             // Note: returns the newly decremented value
             return InterlockedDecrement(&m_value);
 #       elif defined(G3D_LINUX)  || defined(G3D_FREEBSD)
-            unsigned char nz;
-
-            asm volatile ("lock; decl %1;\n\t"
-                          "setnz %%al"
-                          : "=a" (nz)
-                          : "m" (m_value)
-                          : "memory", "cc");
-            return nz;
+            return __sync_sub_and_fetch(&m_value, 1) != 0;
 #       elif defined(G3D_OSX)
             // Note: returns the newly decremented value
             return OSAtomicDecrement32(&m_value);
@@ -143,14 +132,7 @@ public:
 #       if defined(G3D_WINDOWS)
             return InterlockedCompareExchange(&m_value, exchange, comperand);
 #       elif defined(G3D_LINUX) || defined(G3D_FREEBSD) || defined(G3D_OSX)
-            // Based on Apache Portable Runtime
-            // http://koders.com/c/fid3B6631EE94542CDBAA03E822CA780CBA1B024822.aspx
-            int32 ret;
-            asm volatile ("lock; cmpxchgl %1, %2"
-                          : "=a" (ret)
-                          : "r" (exchange), "m" (m_value), "0"(comperand)
-                          : "memory", "cc");
-            return ret;
+            return __sync_val_compare_and_swap(&m_value, comperand, exchange);
 
             // Note that OSAtomicCompareAndSwap32 does not return a useful value for us
             // so it can't satisfy the cmpxchgl contract.
diff --git a/dep/g3dlite/include/G3D/System.h b/dep/g3dlite/include/G3D/System.h
index 4624dd9..50064cc 100644
--- a/dep/g3dlite/include/G3D/System.h
+++ b/dep/g3dlite/include/G3D/System.h
@@ -21,6 +21,10 @@
 #include "G3D/FileNotFound.h"
 #include <string>
 
+#if !defined(_WIN32)
+#include <sys/time.h>
+#endif
+
 #ifdef G3D_OSX
 #define Zone OSX_Zone
 #   include <CoreServices/CoreServices.h>
@@ -494,6 +498,15 @@ public:
 
 #   endif
 
+#elif defined(G3D_LINUX) && defined(__aarch64__)
+
+    // arm64's virtual counter, readable by programs (rdtsc is x86's).
+    inline uint64 System::getCycleCount() {
+        uint64 value;
+        __asm__ __volatile__ ("mrs %0, cntvct_el0" : "=r" (value));
+        return value;
+    }
+
 #elif defined(G3D_LINUX)
 
     inline uint64 System::getCycleCount() {
diff --git a/dep/g3dlite/source/FileSystem.cpp b/dep/g3dlite/source/FileSystem.cpp
index 32a84e7..8e2ae41 100644
--- a/dep/g3dlite/source/FileSystem.cpp
+++ b/dep/g3dlite/source/FileSystem.cpp
@@ -33,6 +33,10 @@
 #else
 #   include <dirent.h>
 #   include <fnmatch.h>
+    // musl's stat is 64-bit already and it has no stat64 (glibc's large-file names).
+#   if !defined(__GLIBC__) && !defined(__APPLE__)
+#       define stat64 stat
+#   endif
 #   include <unistd.h>
 #   define _getcwd getcwd
 #   define _stat stat
diff --git a/dep/g3dlite/source/System.cpp b/dep/g3dlite/source/System.cpp
index 4a0a735..a7a59d7 100644
--- a/dep/g3dlite/source/System.cpp
+++ b/dep/g3dlite/source/System.cpp
@@ -77,7 +77,9 @@
 #endif
 
 // SIMM include
+#if defined(__i386__) || defined(__x86_64__) || defined(_M_IX86) || defined(_M_X64)
 #include <xmmintrin.h>
+#endif
 
 
 namespace G3D {
@@ -1701,7 +1703,11 @@ void System::cpuid(CPUIDFunction func, uint32& eax, uint32& ebx, uint32& ecx, ui
 // for a discussion of why the second version saves ebx; it allows 32-bit code to compile with the -fPIC option.
 // On 64-bit x86, PIC code has a dedicated rip register for PIC so there is no ebx conflict.
 void System::cpuid(CPUIDFunction func, uint32& eax, uint32& ebx, uint32& ecx, uint32& edx) {
-#if ! defined(__PIC__) || defined(__x86_64__)
+#if !defined(__i386__) && !defined(__x86_64__)
+    // No cpuid off x86 (arm64): no vendor, no features.
+    (void)func;
+    eax = ebx = ecx = edx = 0;
+#elif ! defined(__PIC__) || defined(__x86_64__)
     // AT&T assembler syntax
     asm volatile(
                  "movl $0, %%ecx   \n\n" /* Wipe ecx */
